File: backend/app/hashing.py
"""해시 계산과 해시에 들어가는 입력 검사. 규칙의 정본은 docs/HASHING.md (규칙 v1) 이고, 테스트는 docs/hashing_vectors.json 을 읽는다.

저장 시점에 prepare_field / prepare_text 로 한 번만 다듬어 DB 에 넣고, 이후 해시는 그 정본 값으로 계산한다.
"""
import hashlib
import re
import unicodedata
from datetime import date, datetime, timedelta, timezone
from typing import Optional

HASH_RULE_VERSION = 1
ZERO_BYTES32 = "0x" + "0" * 64

US = "\x1f"  # 구분자 U+001F. 파이프가 아니다 (§1)
TRIM = " "  # meta_hash 텍스트 필드는 U+0020 만 지운다 (§1.1)
TRIM_TEXT = " \t\n"  # 여러 줄 텍스트는 U+0020·U+0009·U+000A 를 지운다 (§3)
KST = timezone(timedelta(hours=9))  # 서머타임이 없어 고정 오프셋으로 충분하다. Windows 에 tz 데이터가 없어도 된다

_INVISIBLE = "\u00a0\u200b\u3000\ufeff"  # NBSP, ZWSP, 전각공백, BOM (§1.1)
_FIELD_REJECT = re.compile(f"[\x00-\x1f{_INVISIBLE}]")
_TEXT_REJECT = re.compile(f"[\x00-\x08\x0b-\x1f{_INVISIBLE}]")  # 탭·LF 만 허용 (§3)
_LOWER_BYTES32 = re.compile(r"0x[0-9a-f]{64}")
KST_MIDNIGHT_REMAINDER = 54000  # KST 00:00 = UTC 15:00 → Unix 초 % 86400 (§1.3)


# ---------------------------------------------------------------- 정본 문자열


def canonical(s: str) -> str:
    """meta_hash 텍스트 필드(상호·목적)의 정본. 앞뒤 U+0020 만 지우고 NFC (§1.1)."""
    return unicodedata.normalize("NFC", s.strip(TRIM))


def canonical_text(s: str) -> str:
    """여러 줄 텍스트(사유·이의 본문)의 정본. 개행을 LF 로, 앞뒤 공백·탭·개행 제거, NFC (§3)."""
    s = s.replace("\r\n", "\n").replace("\r", "\n")
    return unicodedata.normalize("NFC", s.strip(TRIM_TEXT))


def prepare_field(value: str, field: str) -> str:
    """상호·목적을 저장 전에 검사하고 정본으로 만든다. 제어문자·보이지 않는 공백·빈 값이면 ValueError (§5)."""
    bad = _FIELD_REJECT.search(value)
    if bad:
        raise ValueError(f"{field}: 허용하지 않는 문자 U+{ord(bad.group()):04X}")
    result = canonical(value)
    if not result:
        raise ValueError(f"{field}: 비어 있다")
    return result


def prepare_text(value: Optional[str], field: str, *, required: bool) -> Optional[str]:
    """사유·이의 본문을 저장 전에 검사하고 정본으로 만든다. 다듬은 뒤 비면 None, required 면 ValueError (§3, §5).

    순서를 지킨다 — 개행 통일을 제어문자 검사보다 먼저 해야 CR 이 섞인 정상 입력이 거부되지 않는다.
    """
    if value is not None:
        value = value.replace("\r\n", "\n").replace("\r", "\n")
        bad = _TEXT_REJECT.search(value)
        if bad:
            raise ValueError(f"{field}: 허용하지 않는 문자 U+{ord(bad.group()):04X}")
        value = canonical_text(value) or None
    if value is None and required:
        raise ValueError(f"{field}: 필수다")
    return value


# ---------------------------------------------------------------- 해시


def _int(v: object, field: str) -> str:
    """정수만 받는다. ORM 이 Decimal·datetime 을 넘겨도 조용히 다른 preimage 가 되지 않게 한다 (§6)."""
    if isinstance(v, bool) or not isinstance(v, int):
        raise TypeError(f"{field}: int 여야 한다 (받은 타입 {type(v).__name__})")
    return str(v)


def meta_preimage(amount: int, counterparty: str, purpose: str, occurred_at: int, receipt_hash: Optional[str]) -> bytes:
    return US.join(
        [_int(amount, "amount"), counterparty, purpose, _int(occurred_at, "occurred_at"), receipt_hash or ""]
    ).encode("utf-8")


def meta_hash(amount: int, counterparty: str, purpose: str, occurred_at: int, receipt_hash: Optional[str]) -> str:
    """counterparty·purpose 는 이미 정본 값이어야 한다. 여기서 다시 다듬지 않는다."""
    return "0x" + hashlib.sha256(meta_preimage(amount, counterparty, purpose, occurred_at, receipt_hash)).hexdigest()


def text_hash(text: str) -> str:
    """text 는 canonical_text 를 거친 정본 값이어야 한다."""
    return "0x" + hashlib.sha256(text.encode("utf-8")).hexdigest()


def reason_hash(text: Optional[str]) -> str:
    """사유가 없으면 bytes32(0). 빈 문자열을 해시하지 않는다 (§3)."""
    return text_hash(text) if text else ZERO_BYTES32


def file_hash(data: bytes) -> str:
    """파일은 바이트 그대로 (§4)."""
    return "0x" + hashlib.sha256(data).hexdigest()


# ---------------------------------------------------------------- 값 검사


def kst_midnight(day: date) -> int:
    """사용일을 KST 자정 Unix 초로 (§1.3)."""
    return int(datetime(day.year, day.month, day.day, tzinfo=KST).timestamp())


def check_occurred_at(ts: int) -> int:
    """사용일의 KST 자정 Unix 초인지 (§1.3). 음수는 % 연산으로 자정처럼 보일 수 있어 따로 막는다 (-32400 % 86400 == 54000)."""
    if ts < 0 or ts % 86400 != KST_MIDNIGHT_REMAINDER:
        raise ValueError("occurred_at: 사용일의 KST 자정이어야 한다 (HASHING §1.3)")
    return ts


def check_bytes32(value: str, field: str = "해시") -> str:
    """0x + 소문자 hex 64자 (§5). 대문자를 고쳐 주지 않는다 — 앱이 서명한 해시와 값이 갈린다 (§4)."""
    if not _LOWER_BYTES32.fullmatch(value):
        raise ValueError(f"{field}: 0x + 소문자 hex 64자여야 한다 (docs/HASHING.md §5)")
    return value


def check_receipt_hash(value: Optional[str]) -> Optional[str]:
    """check_bytes32 와 같고 None(영수증 없음)을 받는다."""
    return None if value is None else check_bytes32(value, "receipt_hash")

File: app/lib/core/hashing.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:unorm_dart/unorm_dart.dart' as unorm;

/// 해시 계산 — `docs/HASHING.md` 규칙 v1 의 Dart 구현
///
/// **정본은 `docs/HASHING.md` 이고 테스트 값은 `docs/hashing_vectors.json` 이다.**
/// 이 파일을 고치기 전에 그 문서를 먼저 볼 것. 두 문서는 `.github/CODEOWNERS`
/// 대상이라 규칙 변경에는 손종인 승인이 필요하다.
///
/// 한 바이트만 달라도 학생 화면의 검증 배지가 전부 빨강으로 뜬다.
///
/// ---
/// **학생 앱에서 [canonical] 을 호출하지 말 것.**
/// 텍스트는 백엔드가 저장 시점에 한 번만 다듬고 그 값이 정본이다 (§1.1).
/// API 가 내려준 값을 앱이 다시 다듬으면 오히려 값이 갈린다.
/// [canonical] 과 [canonicalText] 는 총무·감사 앱이 **서명용 해시를 직접 만들 때**
/// 쓰라고 둔 것이다.
class Hashing {
  Hashing._();

  /// 필드 구분자 — U+001F (Unit Separator). **파이프(`|`)가 아니다.**
  ///
  /// 목적란은 자유 입력이라 총무가 `|` 를 칠 수 있고, 그러면 서로 다른 거래가
  /// 같은 해시를 낸다. `상호="한결문구", 목적="명찰|필기구"` 와
  /// `상호="한결문구|명찰", 목적="필기구"` 가 파이프 방식에서는 preimage 가
  /// 완전히 같아진다. U+001F 는 키보드로 칠 수 없어 이 충돌이 구조적으로 없다.
  static const String unitSeparator = '\u001F';

  /// 규칙 버전. 규칙이 바뀌면 이미 확정된 항목의 배지가 전부 깨지므로,
  /// 버전을 올리기 전에 `HASHING.md` §7 을 먼저 정리해야 한다.
  static const int ruleVersion = 1;

  // ── §1 meta_hash ───────────────────────────────────────────

  /// `meta_hash = SHA256( amount ␟ counterparty ␟ purpose ␟ occurred_at ␟ receipt_hash )`
  ///
  /// [counterparty] 와 [purpose] 는 **이미 정본인 값**을 넘겨야 한다.
  /// 학생 앱은 API 응답을 그대로 넘긴다 (§1.1).
  ///
  /// [receiptHash] 가 null 이면 빈 문자열로 들어가며, **구분자는 그대로 둔다** —
  /// preimage 가 `␟` 로 끝난다.
  static String metaHash({
    required int amount,
    required String counterparty,
    required String purpose,
    required int occurredAt,
    String? receiptHash,
  }) {
    final pre = [
      '$amount',
      counterparty,
      purpose,
      '$occurredAt',
      receiptHash ?? '',
    ].join(unitSeparator);
    return '0x${sha256.convert(utf8.encode(pre))}';
  }

  /// 해시 대상 문자열. 디버깅과 벡터 대조용 — preimage 를 눈으로 볼 때 쓴다.
  static String metaPreimage({
    required int amount,
    required String counterparty,
    required String purpose,
    required int occurredAt,
    String? receiptHash,
  }) {
    return [
      '$amount',
      counterparty,
      purpose,
      '$occurredAt',
      receiptHash ?? '',
    ].join(unitSeparator);
  }

  // ── §3 텍스트 해시 ─────────────────────────────────────────

  /// 사유·이의 본문 등 한 필드짜리 텍스트의 해시.
  ///
  /// 인자는 [canonicalText] 를 거친 정본 값이어야 한다.
  /// 값이 없으면 **빈 문자열을 해시하지 말고** `bytes32(0)` 을 쓴다 (§3).
  static String textHash(String text) =>
      '0x${sha256.convert(utf8.encode(text))}';

  // ── §4 파일 해시 ───────────────────────────────────────────

  /// 파일은 **바이트 그대로** 해시한다. 정규화·트림을 하지 않는다.
  ///
  /// 영수증을 내려받아 이 값을 다시 계산하면, `receipt_hash` 가 실제로
  /// 그 파일의 것인지 확인할 수 있다. 이 단계를 빠뜨리면 영수증만
  /// 바꿔치기한 위조를 못 잡는다 (§2 절차 3번).
  static String fileHash(List<int> bytes) => '0x${sha256.convert(bytes)}';

  // ── §1.1 정본화 — 학생 앱은 호출하지 않는다 ────────────────

  /// `meta_hash` 에 들어가는 한 줄 텍스트의 정본화 (§1.1).
  ///
  /// **제거하는 것은 앞뒤의 U+0020 뿐이다.**
  /// `String.trim()` 을 쓰면 안 된다 — Dart 의 `trim()` 은 U+FEFF(BOM)를 지우는데
  /// Python 의 `strip()` 은 남긴다. 양쪽이 갈리면 해시가 깨진다.
  /// NBSP·전각공백·탭도 남겨야 한다 (`hashing_vectors.json` 의 `canonical_identity` 벡터).
  static String canonical(String s) {
    var t = s;
    while (t.startsWith(' ')) {
      t = t.substring(1);
    }
    while (t.endsWith(' ')) {
      t = t.substring(0, t.length - 1);
    }
    return unorm.nfc(t);
  }

  /// 여러 줄 텍스트의 정본화 (§3).
  ///
  /// 처리 순서를 지킬 것. 제어문자 검사를 개행 변환보다 먼저 하면
  /// CR 이 먼저 걸려 멀쩡한 여러 줄 입력이 전부 거부된다.
  ///   1. CRLF·CR → LF
  ///   2. 제어문자 검사 (백엔드 담당)
  ///   3. 앞뒤 trim (U+0020 · U+0009 · U+000A)
  ///   4. NFC 정규화
  static String canonicalText(String s) {
    var t = s.replaceAll('\r\n', '\n').replaceAll('\r', '\n');
    bool edge(String c) => c == ' ' || c == '\t' || c == '\n';
    while (t.isNotEmpty && edge(t[0])) {
      t = t.substring(1);
    }
    while (t.isNotEmpty && edge(t[t.length - 1])) {
      t = t.substring(0, t.length - 1);
    }
    return unorm.nfc(t);
  }

  // ── 비교 보조 ──────────────────────────────────────────────

  /// 두 해시가 같은 값인지 본다.
  ///
  /// `0x` 접두사 유무와 대소문자는 **표기 차이일 뿐 변조가 아니므로** 맞춰서 본다.
  /// 규칙상 양쪽 모두 `0x` + 소문자로 와야 하지만, 한쪽이 표기를 어겼다고
  /// 「변조 감지」를 띄우면 오탐이 된다.
  ///
  /// 여기서도 `String.trim()` 은 쓰지 않는다 — 해시 문자열에 공백이 붙어 있다면
  /// 그것 자체가 규격 위반이므로 조용히 지워서 통과시키면 안 된다.
  static bool hashEquals(String a, String b) => _bare(a) == _bare(b);

  static String _bare(String hash) {
    final lower = hash.toLowerCase();
    return lower.startsWith('0x') ? lower.substring(2) : lower;
  }

  // ── §1.3 occurred_at ──────────────────────────────────────

  /// KST 자정을 나타내는 Unix 초인지 검사한다.
  ///
  /// 백엔드는 `ts % 86400 == 54000` 이 아니면 400 으로 거부한다.
  /// UTC 자정으로 계산하면 9시간이 어긋나 **전날로 밀린다.**
  static bool isKstMidnight(int epochSeconds) =>
      epochSeconds % 86400 == _kstMidnightRemainder;

  /// KST(UTC+9) 자정의 86400 나머지. `24:00 - 09:00 = 15:00 = 54000초`
  static const int _kstMidnightRemainder = 54000;

  /// 사용일(연·월·일)을 KST 자정 Unix 초로 바꾼다.
  ///
  /// 시·분·초는 버린다. 같은 날이면 언제 등록하든 같은 값이 나와야 한다.
  static int kstMidnightOf(int year, int month, int day) {
    final utc = DateTime.utc(year, month, day);
    // KST 자정은 같은 날 UTC 00:00 보다 9시간 이르다.
    return utc.millisecondsSinceEpoch ~/ 1000 - 9 * 3600;
  }

  /// KST 자정 Unix 초를 사람이 읽는 날짜로 되돌린다.
  static DateTime kstDateOf(int epochSeconds) =>
      DateTime.fromMillisecondsSinceEpoch(
        (epochSeconds + 9 * 3600) * 1000,
        isUtc: true,
      );
}
